R/clean_redcap_long.R

```diff
@@ -337,3 +349,95 @@ distill_repeat_table_long <- function(form_name,
   out %>%
     tibble()
 }
+
+#' @title Convert Mixed Structure Instruments to Repeating Instruments
+#'
+#' @description
+#' For longitudinal projects where users set `enable_mixed_structure` to `TRUE`,
+#' this function will handle the process of setting the nonrepeating parts of the
+#' instrument to repeating ones with a single instance.
+#'
+#' @param db_data_long The longitudinal REDCap database output defined by
+#' \code{REDCapR::redcap_read_oneshot()$data}
+#' @param db_metadata_long The longitudinal REDCap metadata output defined by
+#' \code{REDCapR::redcap_metadata_read()$data}
+#'
+#' @return
+#' Returns a \code{tibble} with list elements containing tidy dataframes. Users
+#' can access dataframes under the \code{redcap_data} column with reference to
+#' \code{form_name} and \code{structure} column details.
+#'
+#' @keywords internal
+
+convert_mixed_instrument <- function(db_data_long, db_metadata_long) {
```

Comment:
We're losing data somewhere in here. When we convert the rows to have a value for `redcap_repeat_instrument` that was previously `NA` I think we're ending up throwing away the nonrepeating data from that row.

Reply:
Another structure to consider that makes me wonder if our current approach actually generalizes:

|Event|Type|Nonrepeating Instruments|Repeating Instruments|
|-|-|-|-|
|Event 1 | Not Repeating | inst_1, inst_2, inst_3 | N/A|
|Event 2 | Repeat Instruments | inst_1 | inst_2, inst_3|

For the row in the data from `Event 1`, how would we be assigning `redcap_repeat_instrument`? In the test redcap it's easy because there's only data from one repeating instrument there but here we have data from 2 different repeating instruments.

Reply:
> We're losing data somewhere in here. When we convert the rows to have a value for redcap_repeat_instrument that was previously NA I think we're ending up throwing away the nonrepeating data from that row.

Ah, yes good catch. Will work on this one, seems less trivial than I was hoping. I'm not sure I follow the second comment but maybe fixing the first will address it.

Reply:
Yeah I'd say address the first thing and then once that's working I can mess with the test redcap a bit to see if I can make the second thing I was saying more concrete

Reply:
First part addressed, let me know if this helps address the second.

Reply:
I made a new redcap that shows the issue. You have access to poke around but feel free to just use my token below. We'll delete this one later.

```r
devtools::load_all()

read_redcap(
  Sys.getenv("REDCAP_URI"),
  "0AE465FB49C411385DC8FE1C5A9638C0",
  allow_mixed_structure = TRUE
) |>
  extract_tibble("mixed_structure_form_2")
```

This one has 2 mixed structure instruments. The issue is that one row of the API export (the first row in this case) contains data from both mixed structure instruments. In `convert_mixed_instrument()` we assign a single instrument for each row of the export. It ends up just being the first one it sees. That results in the nonrepeating data from the second mixed instrument getting lost ("Mixed 2 Nonrepeat 1").

I haven't quite wrapped my head around this enough to suggest a fix but we're going to need to make things flexible enough that single rows of the export can get mapped to multiple repeating instruments.

Reply:
Ugh, alright I also need to wrap my head around this more. Thanks for the catch and explanation.

Reply:
Most recent commit: I _think_ this works (it seems to resolve your REDCap and maintain the old one, all tests and checks pass). 

The idea is that by moving the `convert_mixed_instrument()` function into the `distill_*()` function, we update `redcap_repeat_instrument` with each evaluation of the `map()` in `clean_redcap_long()`...

If I could post a GIF from Inception I would here. Before CRAN submission I almost want to ask the folks who submitted issues to test it.

Reply:
Ah clever! That sounds like it makes sense but I'll take a closer look. Agree that once we get this merged we should ask the issue submitters to try out the dev version before submitting to CRAN